### PR TITLE
maxim changes

### PR DESCRIPTION
**Pull request by danilovmy**
i try to make something compact
to see an original - you shoudl kill the site_id_fallback

Don't Made yet : save all languages, save origin.
Save all lang - for lang, val in vars(translator).items(); save Translation. 
Save origin ... I dont decide yet - i can save it in model or in Translation.

### CODE REVIEW

File: tof/models.py [hunk 1]
```diff
@@ -98,26 +75,69 @@ class Meta:
...
+        setattr(translation, translation.get_lang() if hasattr(instance, '_end_init') else '_origin', str(value))
 
-        restore_cls_after_translate(
-            cls,
-            name,
-            ContentType.objects.filter(translatablefield__isnull=False, id=ct_pk).exists(),
-        )
+    def __delete__(self, instance):
+        vars(self).pop(self.name, None)
+        del type(instance)._meta._field_tof[self.attname]  # pragma: no cover
```
Comment (mom1) [inline tof/models.py]: self.attname shoul be self.name

File: tof/models.py [hunk 2]
```diff
@@ -98,26 +75,69 @@ class Meta:
...
+
+    def save_translation(self, instance):
+        val = instance.get_translation(self.name)
+        if val:
+            translation, _ = self.translations.get_or_create(object_id=instance.pk, lang_id=val.get_lang(),)
```
Comment (mom1) [inline tof/models.py]: It do not work. Need pass content_type_id=self.content_type_id

File: tof/models.py [hunk 3]
```diff
@@ -98,26 +75,69 @@ class Meta:
...
 
     def delete(self, *args, **kwargs):
-        cls = self.content_type.model_class()
-        ct_pk = self.content_type.pk
-        name = self.name
         super().delete(*args, **kwargs)
+        self.remove_translation_from_class()
+
+    def __get__(self, instance, instance_cls):
+        return instance.get_translation(self.name)
```
Comment (mom1) [inline tof/models.py]: instance = None if call Wine.title

File: tof/models.py (hunk 1 above)
Comment (danilovmy) [reply tof/models.py]: fixed

File: tof/models.py (hunk 2 above)
Comment (danilovmy) [reply tof/models.py]: fixed

File: tof/models.py (hunk 3 above)
Comment (danilovmy) [reply tof/models.py]: its right, or i dont understand. what i should return?

Comment (mom1) [reply tof/models.py]: Maybe self?

File: tof/models.py [hunk 4]
```diff
@@ -98,26 +75,69 @@ class Meta:
...
+        setattr(cls, cls._meta._field_tof.setdefault(self.id, self).name, self)
+
+    def remove_translation_from_class(self):
+        cls = self.content_type.model_class()
+        cls._meta._field_tof.pop(self.id)
+        delattr(cls, self.name)
+        field = cls._meta.get_field(self.name)
+        field.contribute_to_class(cls, self.name)
+        if issubclass(cls, TranslationFieldMixin) and not cls._meta._field_tof:
+            cls.__bases__ = cls.__bases__[1:]  # TODO DONT USE [1:], it can be not first!!!!
```
Comment (mom1) [inline tof/models.py]: You missed befor/after this line `del cls._meta._field_tof`

File: tof/models.py [hunk 5]
```diff
@@ -98,26 +75,69 @@ class Meta:
...
+                origin = cls.objects
+                new_mng_cls = type(f'TranslationManager{cls.__name__}', (TranslationManager, type(origin)), {})
+                trans_mng = trans_mng or new_mng_cls()
+                trans_mng.contribute_to_class(cls, trans_mng.default_name)
+                cls._meta.default_manager_name = trans_mng.default_name
+                # FIXME
+                del cls.objects
+                trans_mng.contribute_to_class(cls, 'objects')
+                origin.contribute_to_class(cls, 'objects_origin')
+        setattr(cls, cls._meta._field_tof.setdefault(self.id, self).name, self)
```
Comment (mom1) [inline tof/models.py]: If you change _field_tof key, you must will change [expand_filter decorator](https://github.com/mom1/django-tof/blob/master/tof/decorators.py#L68) else you break filter feature

File: tof/utils.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    @staticmethod
+    def get_lang():
+        lang, *_ = get_language().partition('-')
+        return lang
+
+    def get_fallback_languages(self, attr):
+        for fallback in (FALLBACK_LANGUAGES.get(attr) or (), FALLBACK_LANGUAGES.get(SITE_ID) or ()):
+            # probably i should return isinstance list tuple 
+            yield from (lang for lang in fallback if lang != attr)
+        yield DEFAULT_LANGUAGE
```
Comment (mom1) [inline tof/utils.py]: what about if `FALLBACK_LANGUAGES = {'fr': 'nl'}`
do it not supporting?

Comment (danilovmy) [reply tof/utils.py]: # probably i should return isinstance list tuple

File: tof/models.py (hunk 5 above)
Comment (danilovmy) [reply tof/models.py]: i dont seen yet filter_decorator. but it's right now can be more easy too.

File: tof/utils.py (hunk 6 above)
Comment (mom1) [reply tof/utils.py]: what you mean?

Comment (mom1) [reply tof/utils.py]: like this?
`yield from (lang for lang in fallback if lang != attr) if isinstance(fallback, (list, tuple)) else (i for i in (fallback,))`

File: tof/models.py [hunk 7]
```diff
@@ -98,26 +75,69 @@ class Meta:
...
         super().save(*args, **kwargs)
-        prepare_cls_for_translate(self.content_type.model_class(), self)
+        self.add_translation_to_class()
 
     def delete(self, *args, **kwargs):
-        cls = self.content_type.model_class()
-        ct_pk = self.content_type.pk
-        name = self.name
         super().delete(*args, **kwargs)
+        self.remove_translation_from_class()
```
Comment (mom1) [inline tof/models.py]: It do not work after delete because self.id is None

File: tof/decorators.py [hunk 8]
```diff
@@ -87,6 +84,6 @@ def expand_filter(model_cls, key, value):
         else:
             query &= Q(lang=get_language())
         query &= Q(**{f'value{sep}{lookup}': value})
-        new_val = Translation.objects.filter(query).values_list('object_id', flat=True)
+        new_val = field.tarnslations.filter(query).values_list('object_id', flat=True)
```
Comment (mom1) [inline tof/decorators.py]: field is str

File: tof/utils.py [hunk 9]
```diff
@@ -37,7 +37,8 @@ def get_lang():
         return lang
 
     def get_fallback_languages(self, attr):
-        for fallback in (FALLBACK_LANGUAGES.get(attr) or (), FALLBACK_LANGUAGES.get(SITE_ID) or ()):
-            # probably i should return isinstance list tuple 
-            yield from (lang for lang in fallback if lang != attr)
-        yield DEFAULT_LANGUAGE
\ No newline at end of file
+        for fallback in (FALLBACK_LANGUAGES.get(attr) or (), FALLBACK_LANGUAGES.get(SITE_ID) or (), DEFAULT_LANGUAGE):
+            if isinctance(fallback, (list, tuple))
```
Comment (mom1) [inline tof/utils.py]: syntax error

File: tof/utils.py [hunk 10]
```diff
@@ -13,14 +13,14 @@ def __init__(self, **kwargs):
...
-    def __getattr__(self, name):
-        if name in ('resolve_expression', 'prepare_database_save'):
-            raise AttributeError(name)
-        attrs = vars(self)
-        for val_lang in self.get_fallback_languages(name):
-            if val_lang in attrs:  # if attr exists we should return value, dont matter what is it.
-                return attrs[val_lang]
-        return attrs.get('_origin') or ''
+    def __getattr__(self, attr):
+        if len(attrs) == 2:
```
Comment (mom1) [inline tof/utils.py]: local variable 'attrs' referenced before assignment

File: tof/models.py [hunk 11]
```diff
@@ -98,26 +75,70 @@ class Meta:
...
+                cls._meta.default_manager_name = trans_mng.default_name
+                # FIXME
+                del cls.objects
+                trans_mng.contribute_to_class(cls, 'objects')
+                origin.contribute_to_class(cls, 'objects_origin')
+        setattr(cls, cls._meta._field_tof.setdefault(self.id, self).name, self)
+
+    def remove_translation_from_class(self):
+        cls = self.content_type.model_class()
+        cls._meta._field_tof.pop(self.id)
```
Comment (mom1) [inline tof/models.py]: KeyError: None 
test: TranslatableFieldTestCase.test_delete

File: tof/models.py [hunk 12]
```diff
@@ -98,26 +75,70 @@ class Meta:
...
 
     def delete(self, *args, **kwargs):
-        cls = self.content_type.model_class()
-        ct_pk = self.content_type.pk
-        name = self.name
         super().delete(*args, **kwargs)
+        self.remove_translation_from_class()
+
+    def __get__(self, instance, instance_cls):
+        return instance.get_translation(self.name)
```
Comment (mom1) [inline tof/models.py]: AttributeError: 'NoneType' object has no attribute 'get_translation'
test: TranslationFieldMixinTestCase.test_get

probably need change test.

File: tof/models.py (hunk 7 above)
Comment (danilovmy) [reply tof/models.py]: fix

File: tof/decorators.py (hunk 8 above)
Comment (danilovmy) [reply tof/decorators.py]: fix

File: tof/utils.py (hunk 9 above)
Comment (danilovmy) [reply tof/utils.py]: fix

File: tof/utils.py (hunk 10 above)
Comment (danilovmy) [reply tof/utils.py]: fix

File: tof/models.py (hunk 11 above)
Comment (danilovmy) [reply tof/models.py]: fix

File: tof/decorators.py [hunk 13]
```diff
@@ -84,6 +84,6 @@ def expand_filter(model_cls, key, value):
         else:
             query &= Q(lang=get_language())
         query &= Q(**{f'value{sep}{lookup}': value})
```
Comment (mom1) [inline tof/decorators.py]: 1. Probably it must befor first if
2. 73 line Always False

File: tof/decorators.py [hunk 14]
```diff
@@ -84,6 +84,6 @@ def expand_filter(model_cls, key, value):
...
-        new_val = field.tarnslations.filter(query).values_list('object_id', flat=True)
+        new_val = field.translations.filter(query).values_list('object_id', flat=True)
```
Comment (mom1) [inline tof/decorators.py]: field still string. Maybe this:
```
field, sep, lookup = key.partition('__')
field = next((f for f in model_cls._meta._field_tof.values() if f.name == field), None)
if field:
    query = Q(**{f'value{sep}{lookup}': value})
    if DEFAULT_FILTER_LANGUAGE == '__all__':
            ...
```

File: tof/models.py [hunk 15]
```diff
@@ -121,7 +121,7 @@ def add_translation_to_class(self, trans_mng=None):
                 del cls.objects
                 trans_mng.contribute_to_class(cls, 'objects')
                 origin.contribute_to_class(cls, 'objects_origin')
-        setattr(cls, cls._meta._field_tof.setdefault(self.id, self).name, self)
+        setattr(cls, cls._meta._field_tof.setdefault(self.id, self), self)
```
[the remark this answers is not in the document]
Comment (danilovmy) [reply tof/models.py]: right!

File: tof/utils.py [hunk 16]
```diff
@@ -38,7 +38,7 @@ def get_lang():
 
     def get_fallback_languages(self, attr):
         for fallback in (FALLBACK_LANGUAGES.get(attr) or (), FALLBACK_LANGUAGES.get(SITE_ID) or (), DEFAULT_LANGUAGE):
-            if isinctance(fallback, (list, tuple))
+            if isinctance(fallback, (list, tuple)):
```
[the remark this answers is not in the document]
Comment (danilovmy) [reply tof/utils.py]: probably

File: tof/utils.py [hunk 17]
```diff
@@ -14,7 +14,7 @@ def __init__(self, **kwargs):
         vars(self).update(**kwargs)
 
     def __getattr__(self, attr):
-        if len(attrs) == 2:
+        if len(attr) == 2:
```
[the remark this answers is not in the document]
Comment (danilovmy) [reply tof/utils.py]: class Language(models.Model):
    iso = models.CharField(max_length=2)
only 2 char is possible

**Comment by mom1** [comment]
merged

**Pull request closed without merging**